### PR TITLE
[casr-afl][casr-libfuzzer] Support timeout

### PR DESCRIPTION
**Pull request by hkctkuy**

### CODE REVIEW

File: casr/src/bin/casr-libfuzzer.rs [hunk 1]
```diff
@@ -170,9 +191,28 @@ fn main() -> Result<()> {
...
             debug!("{:?}", casr_cmd);
-            let casr_output = casr_cmd
-                .output()
-                .with_context(|| format!("Couldn't launch {casr_cmd:?}"))?;
+
+            // Get output
+            let casr_output =
+            // If timeout is specified, spawn and check timeout
+            // Else get output
+            if timeout != 0 {
```
Comment (SweetVishnya) [inline casr/src/bin/casr-libfuzzer.rs:199]: We may move this code to utility function

File: casr/src/bin/casr-libfuzzer.rs [hunk 2]
```diff
@@ -170,9 +191,28 @@ fn main() -> Result<()> {
...
+                    .wait_timeout(Duration::from_secs(timeout))
+                    .unwrap()
+                    .is_none()
+                {
+                    child.kill()?;
+                    warn!("Timeout: {:?}", casr_cmd);
+                }
+                child.wait_with_output()?
+            } else {
+                casr_cmd.output().with_context(|| format!("Couldn't launch {casr_cmd:?}"))?
```
Comment (SweetVishnya) [inline casr/src/bin/casr-libfuzzer.rs:213]: we should support timeout in casr-san, etc. and call it with -t here

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: casr/src/bin/casr-libfuzzer.rs (hunk 1 above)
Comment (hkctkuy) [reply casr/src/bin/casr-libfuzzer.rs]: Done

File: casr/src/bin/casr-libfuzzer.rs (hunk 2 above)
Comment (hkctkuy) [reply casr/src/bin/casr-libfuzzer.rs]: Done (except casr-gdb so far)

Comment (hkctkuy) [reply casr/src/bin/casr-libfuzzer.rs]: Done for casr-gdb too

File: casr/src/bin/casr-san.rs [hunk 3]
```diff
@@ -115,17 +131,15 @@ fn main() -> Result<()> {
...
-            .with_context(|| "Couldn't run target program with sanitizers")?
+    let sanitizers_cmd = unsafe {
+        sanitizers_cmd.pre_exec(|| {
+            if personality(linux_personality::ADDR_NO_RANDOMIZE).is_err() {
+                panic!("Cannot set personality");
+            }
+            Ok(())
+        })
     };
+    let sanitizers_result = util::get_output(sanitizers_cmd, timeout)?;
```
Comment (SweetVishnya) [inline casr/src/bin/casr-san.rs:142]: Also, set timeout for GdbCommand call (line 223)

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: casr/src/bin/casr-afl.rs [hunk 4]
```diff
@@ -32,8 +32,14 @@ impl<'a> AflCrashInfo {
     ///
     /// # Arguments
     ///
-    /// `output_dir` - save report to specified directory or use the same directory as crash
-    pub fn run_casr<T: Into<Option<&'a Path>>>(&self, output_dir: T) -> anyhow::Result<()> {
+    /// * `output_dir` - save report to specified directory or use the same directory as crash
+    ///
+    /// * `timeout` - target program timeout
```
Comment (anfedotoff) [inline casr/src/bin/casr-afl.rs:37]: in seconds?  Update comment then.

**Review by anfedotoff — CHANGES_REQUESTED** [review]

File: casr/src/util.rs [hunk 5]
```diff
@@ -248,3 +251,37 @@ pub fn log_progress(processed_items: &RwLock<usize>, total: usize) {
...
+        let mut child = command
+            .spawn()
+            .with_context(|| "Failed to start command: {command:?}")?;
+        if child
+            .wait_timeout(Duration::from_secs(timeout))
+            .unwrap()
+            .is_none()
+        {
+            let _ = child.kill();
+            warn!("Timeout: {:?}", command);
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/util.rs]: Done

File: casr/src/bin/casr-gdb.rs [hunk 6]
```diff
@@ -167,6 +184,7 @@ fn main() -> Result<()> {
         .siginfo()
         .mappings()
         .regs()
+        .timeout(timeout)
```
[the remark this answers is not in the document]
Comment (hkctkuy) [reply casr/src/bin/casr-gdb.rs]: Done

File: casr/src/bin/casr-afl.rs (hunk 4 above)
Comment (hkctkuy) [reply casr/src/bin/casr-afl.rs]: Done

File: casr/src/util.rs [hunk 7]
```diff
@@ -248,3 +251,43 @@ pub fn log_progress(processed_items: &RwLock<usize>, total: usize) {
...
+            .spawn()
+            .with_context(|| "Failed to start command: {command:?}")?;
+        if child
+            .wait_timeout(Duration::from_secs(timeout))
+            .unwrap()
+            .is_none()
+        {
+            let _ = child.kill();
+            if error_on_timeout {
+                error!("Timeout: {:?}", command);
```
Comment (anfedotoff) [inline casr/src/util.rs:282]: Just use error instead of warn, remove this parameter

**Review by anfedotoff — CHANGES_REQUESTED** [review]

File: casr/src/util.rs (hunk 7 above)
Comment (hkctkuy) [reply casr/src/util.rs]: Replaced with bail

**Review by SweetVishnya — APPROVED** [review]

**Review by anfedotoff — APPROVED** [review]

**Pull request merged**